extensions/amp-base-carousel/1.0/scroller.js

```diff
@@ -0,0 +1,231 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as Preact from '../../../src/preact';
+import * as styles from './base-carousel.css';
+import {WithAmpContext} from '../../../src/preact/context';
+import {debounce} from '../../../src/utils/rate-limit';
+import {mod} from '../../../src/utils/math';
+import {useLayoutEffect, useRef} from '../../../src/preact';
+
+/**
+ * How long to wait prior to resetting the scrolling position after the last
+ * scroll event. Ideally this should be low, so that once the user stops
+ * scrolling, things are immediately centered again. Since there can be some
+ * delay between scroll events, and we do not want to interrupt a scroll with a
+ * render, it cannot be too small. 200ms seems to be around the lower limit for
+ * this value on Android / iOS.
+ */
+const RESET_SCROLL_REFERENCE_POINT_WAIT_MS = 200;
+
+/**
+ * @param {!JsonObject} props
+ * @return {PreactDef.Renderable}
+ */
+export function Scroller(props) {
+  const {
+    children,
+    ignoreProgrammaticScroll,
+    loop,
+    restingIndex,
+    setRestingIndex,
+  } = props;
+
+  /**
+   * The number of slides we want to place before the
+   * reference or resting index. Only needed if loop=true.
+   */
+  const pivotIndex = Math.floor(children.length / 2);
+
+  /**
+   * The dynamic position that the slide at the resting index
+   * is with respect to its scrolling order. Only needed if loop=true.
+   */
+  const offsetRef = useRef(restingIndex);
+  const containerRef = useRef();
+  const slides = renderSlides({
+    children,
+    loop,
+    offsetRef,
+    pivotIndex,
+    restingIndex,
+  });
+  const currentIndex = useRef(restingIndex);
+
+  // Note: this has to be useLayoutEffect, not useEffect.
```

Comment:
Let's just write why. E.g. "Layout effect is used to avoid FOUC with scrolling" or such.